teps/0042-taskrun-breakpoint-on-failure.md

```diff
@@ -0,0 +1,119 @@
+focused.  If you disagree with what is already in a document, open a new PR
+with suggested changes.
+
+If there are new details that belong in the TEP, edit the TEP.  Once a
+feature has become "implemented", major changes should get new TEPs.
+
+The canonical place for the latest set of instructions (and the likely source
+of this file) is [here](/teps/NNNN-TEP-template/README.md).
+
+-->
+
+<!--
+This is the title of your TEP.  Keep it short, simple, and descriptive.  A good
+title can help communicate what the TEP is and should be considered as part of
+any review.
+-->
+
+<!--
+A table of contents is helpful for quickly jumping to sections of a TEP and for
+highlighting any additional information provided beyond the standard TEP
+template.
+
+Ensure the TOC is wrapped with
+  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+If a TaskRun fails in Tekton, it can be hard to understand at what step and what exactly in the step led to the failure.
+To solve this problem, it can be possible for the users to set a breakpoint on failure specification in the TaskRun which
+halts it at a particular step, making it easy for the user to debug it. The following proposal covers the implementation
+details regarding the same.
+
+## Motivation
+
+With the addition of a "breakpoint on failure" mechanism for TaskRuns, the user experience should vastly improve as traditional
+CI/CD systems have only a primitive debugging present, whereas while leveraging containers we can make Live Debugging in
+Tekton possible with this.
+
+The user would enjoy 
+
+- Improved Experience
+- Faster iteration during Pipeline development
+- Real time debugging 
+
+### Goals
+
+When the "breakpoint on failure" spec is mentioned in a particular TaskRun it should
+- Pause TaskRun on failure of a step
+- Allow User to get shell access to a debugging environment
+- Manually continue the TaskRun from where the breakpoint occurred
+
+### Non-Goals
+
+- PipelineRun debug/breakpoint
+- Breakpoint on other steps apart from the failing one (not in scope)
+
```

Comment:
hey @waveywaves I know you recently demo'ed this

Does it make sense to 
a) post a URL to the demo that is accessible at least to this TEP's reviewers while we wait on further details here?
b) or are those details coming soon and we should wait for them, 
c) or a) and b) :-) 

thanks

Reply:
Here is a demo showcasing TaskRun Breakpoint on failure
[![asciicast](https://asciinema.org/a/383256.svg)](https://asciinema.org/a/383256)

Reply:
Great demo! This looks very useful. I'd just echo Christie's comment below about fleshing out the users stories and requirements. From my perspective, the most useful use case would be something along the lines of I pick a step within a task to halt on, and the debug framework halts execution at the beginning of the step. Then I can exec in and play around.